Components: Add a WAI-ARIA compliant custom select.

epiqueras opened the pull request:
Closes #16473

## Description

This PR implements a fully accessible and customizable select component with individually style-able items, in the `@wordpress/components` package.

It uses [Downshift](https://github.com/downshift-js/downshift), a library that will be very useful for any of our other complex a11y input needs.

It has some modifications hooked on top of it to make it fully compliant with [the spec](https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html). These modifications might soon make it into a future [Downshift](https://github.com/downshift-js/downshift) release and at that point we should remove them from our implementations.

**The story for the component is:**

https://deploy-preview-17926--gutenberg-playground.netlify.com/design-system/components/?path=/story/customselect--default

## How has this been tested?

It was verified that the component works as expected.

Although the libraries used are heavily tested, unit testing this implementation once the approach is approved wouldn't be a bad idea to make sure we never regress to configuring something incorrectly.

## Screenshots

<img width="789" alt="Screen Shot 2019-10-13 at 12 46 18 PM" src="https://user-images.githubusercontent.com/19157096/66721104-79d04080-edb7-11e9-8e2c-5563098ef779.png">

## Types of Changes

*New Feature:* `@wordpress/components` now has a fully WAI-ARIA compliant custom select.

## Checklist:

- [x] My code is tested.
- [x] My code follows the WordPress code style. <!-- Check code: `npm run lint`, Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/javascript/ -->
- [x] My code follows the accessibility standards. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/coding-standards/accessibility-coding-standards/ -->
- [x] My code has proper inline documentation. <!-- Guidelines: https://make.wordpress.org/core/handbook/best-practices/inline-documentation-standards/javascript/ -->
- [x] I've included developer documentation if appropriate. <!-- Handbook: https://developer.wordpress.org/block-editor/ -->

epiqueras committed:
Components: Add a WAI-ARIA compliant custom select.

silviuavram commented on packages/components/src/custom-select/index.js:
```diff
@@ -0,0 +1,123 @@
+	{ type, changes, props: { items } }
+) => {
+	switch ( type ) {
+		case Downshift.stateChangeTypes.ToggleButtonKeyDownArrowDown:
```
it's actually `useSelect.stateChangeTypes` here and below.

epiqueras replied:
Thank you!

silviuavram commented on packages/components/src/custom-select/index.js:
```diff
@@ -0,0 +1,123 @@
+// the menu does not necessarily open on
+// key up/down, you can still switch between
+// options with the menu closed.
+const stateReducer = (
```
related to the Dragon Speech issue, I can look into it. if you are faster at it, I would suggest adding a console log here in the stateReducer and check the `type`, `changes` and `state`. If there is a state change in the select, then it should pass here.

epiqueras replied:
Great idea!

youknowriad commented:
Design-wise the options should be shown in a Popover I think.

epiqueras commented:
>Design-wise the options should be shown in a Popover I think.

I want to tick all the `a11y` boxes first, and then see what we can do with styling without regressing.

epiqueras committed:
Custom Select: Fix Downshift constant references.

epiqueras committed:
Custom Select: Add explicit aria-label to button and export component.

epiqueras committed:
Custom Select: Add temporary console debugging.

tellthemachines commented on packages/components/src/custom-select/index.js:
```diff
@@ -0,0 +1,137 @@
+import { Button, Dashicon } from '../';
+
+const itemToString = ( item ) => item.name;
+// This is needed so that in Windows, where
```
Just checked this on Windows and the behaviour is identical to Mac: on arrow down/up the menu opens. I don't see that as a major issue though. If we did want the arrow keys to work with the menu closed we'd probably have to make sure the options always render, and hide them with CSS, but then it's not clear that we could limit that behaviour to Windows 🤷‍♀

silviuavram replied:
the issue here is the returned `changes`. @epiqueras just return the `selectedItem` because if you also return the rest of the `changes` then one of them is `isOpen` that comes as true :D

epiqueras replied:
Fixed :smile:

tellthemachines commented on packages/components/src/custom-select/index.js:
```diff
@@ -0,0 +1,137 @@
+				{ ...getLabelProps( { className: 'components-custom-select__label' } ) }
+			>
+				{ label }
+			</span>
```
Any reason not to make this an actual `<label>` instead of using `aria-labelledby`? Then we might not need the `aria-label`. Having both, VoiceOver and NVDA are announcing the button as "Font size font size". Would be good to avoid the repetition.

silviuavram replied:
The way I wrote `useSelect` was with having a physical label that also labels the button and the menu. As I understood there was an issue with Dragon Speech because of this.

Having said that, anyone tried Dragonn Speech to select an item then check what the console.log prints?

epiqueras replied:
>Any reason not to make this an actual <label> instead of using aria-labelledby? Then we might not need the aria-label. Having both, VoiceOver and NVDA are announcing the button as "Font size font size". Would be good to avoid the repetition.

Dragon Speech still needs the explicit aria label on the toggle to be able to click on it.

>Having said that, anyone tried Dragonn Speech to select an item then check what the console.log prints?

It registers a click on the toggle button I think:

<img width="815" alt="Screen Shot 2019-10-18 at 1 58 17 PM" src="https://user-images.githubusercontent.com/19157096/67128001-5fc7a100-f1af-11e9-9900-a70da5d4d519.png">

In that test^^ "Small" was already the selected item.

epiqueras replied:
Done!

tellthemachines reviewed (commented):
Good work on this! Checked with VO/Safari and NVDA/Firefox and it's behaving pretty well on both, apart from the repetition of "Font size" label I commented on below.

One thing I noticed on NVDA but was unable to reproduce consistently is that sometimes the font size button is announced as "expanded" when focused, and with the dropdown closed. I realise this may not be very helpful, will try and debug further 😄 

Another potential issue is that focus is not being limited to the component when it is open, so if we open the dropdown and then press tab we move out of it but the dropdown stays open. Shift-tabbing until the font size button is focused again allows us to continue interacting with the dropdown, but after that pressing `esc` will not close it. 

Also, minor styling issue on Firefox and IE:

<img width="164" alt="Screen Shot 2019-10-18 at 11 36 46 am" src="https://user-images.githubusercontent.com/8096000/67057447-42d5a280-f19c-11e9-900d-1a519ae48075.png">

Removing `display: inline-flex` and adding `text-align: left` to `components-button` will fix it for both browsers 🙂

epiqueras committed:
Custom Select: Collapsed switching fix.

epiqueras committed:
Custom Select: Firefox/IE styling fix.

epiqueras commented:
@tellthemachines Thanks for the help!

>Removing display: inline-flex and adding text-align: left to components-button will fix it for both browsers 🙂

Done :smile:

tellthemachines reviewed (changes requested):
We're getting there! 

One remaining issue, apart from the comments I left: pressing Tab when the dropdown is open still causes focus to be moved to the button control. Subsequently pressing Esc will _not_ close the dropdown, but pressing any arrow key _will_ close it.

Also, would be good to test replacing the current font size picker with this component, to check focus management when entering/exiting it.

enriquesanchez commented:
I ran some tests of the component in https://deploy-preview-17926--gutenberg-playground.netlify.com/design-system/components/?path=/story/customselect--default with Dragon, NVDA and Voice Over.

Here are the videos of the test:

**With Dragon: https://cldup.com/cihauaUTN4.m4v**
So far with Dragon I haven't able to select an option from the list by saying "Click Large", but I'm able to do so if I use the "move up/down" commands to switch focus to an option and then say "press enter".  I'm a fairly new user of speech recognition software, so I'm not sure how common or uncommon this flow is. That being said, I was able to interact with the component and select an option.

**With NVDA: https://cldup.com/7n79VbZMo3.m4v** (sorry for the shaky camera)
Besides "Font Size" being announced repetitively, the only thing that I felt was missing was that on Windows I can't change the selection by using the up/down keys, like on Mac.

**With Voice Over: https://cldup.com/A12FmcNrSn.m4v**
Same as with NVDA, "Font Size" is announced repetitively. Othe than that, interaction was as expected.

tellthemachines commented:
> Besides "Font Size" being announced repetitively, the only thing that I felt was missing was that on Windows I can't change the selection by using the up/down keys, like on Mac.

@enriquesanchez this happens because NVDA is not switching into forms mode when entering this component. If you manually switch by pressing Insert+Space then you'll be able to use up/down keys. It would be worth investigating if it's possible to make NVDA switch automatically as that would be closer in behaviour to a native control.

adamsilverstein commented:
Wanted to mention previous work to use Downshift for the post author selector in https://github.com/WordPress/gutenberg/pull/7478. It might be worth reviewing the comments on that ticket - ultimately, the accessibility team rejected this approach and we settled instead on using `accessible-autocomplete` - see https://github.com/WordPress/gutenberg/pull/7385.

@enriquesanchez - had you reviewed this previous work? What do you think of Downshift vs. accessible-autocomplete?

silviuavram commented:
Adam can you mention the reasons for making that decision? I am interested
for improvement reasons at least. Thanks!

On Sun, 3 Nov 2019 at 15:30, Adam Silverstein <notifications@github.com>
wrote:

> Wanted to mention previous work to use Downshift for the post author
> selector in #7478 <https://github.com/WordPress/gutenberg/pull/7478>. It
> might be worth reviewing the comments on that ticket - ultimately, the
> accessibility team rejected this approach and we settled instead on using
> accessible-autocomplete - see #7385
> <https://github.com/WordPress/gutenberg/pull/7385>.
>
> @enriquesanchez <https://github.com/enriquesanchez> - had you reviewed
> this previous work? What do you think of Downshift vs.
> accessible-autocomplete?
>
> —
> You are receiving this because you commented.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=ACWAZACSHLQ5J4LF4BSAP3TQR3N7PA5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEC5VMIQ#issuecomment-549148194>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWAZAHD6WIVNVM6P4PRGDTQR3N7PANCNFSM4JAIMOLQ>
> .
>

epiqueras committed:
Font Size Picker: Replace `SelectControl` with `CustomSelect`.

epiqueras commented on packages/components/src/custom-select/style.scss:
```diff
@@ -0,0 +1,50 @@
+	border: 1px solid $dark-gray-200;
+	border-radius: 4px;
+	color: $dark-gray-500;
+	display: initial;
```
Done!

epiqueras committed:
Custom Select: A11y review.

epiqueras commented:
I've made some optimizations and replaced the font size picker's select with this one.

Can we move forward with this PR and iterate in following ones now?

@tellthemachines @enriquesanchez @youknowriad

tellthemachines commented:
> Adam can you mention the reasons for making that decision? I am interested for improvement reasons at least. Thanks!

@silviuavram [this comment](https://github.com/WordPress/gutenberg/pull/7478#issuecomment-400599022) explains the reasons, though that was over a year ago so perhaps some of those issues have been fixed by now.

tellthemachines commented on packages/components/src/custom-select/index.js:
```diff
@@ -0,0 +1,148 @@
+			<Button
+				{ ...getToggleButtonProps( {
+					// This is needed because some speech recognition software don't support `aria-labelledby`.
+					'aria-label': label,
```
The aria-label on the button will need to be the same as the visible label, because that's what speech to text users will try to activate the button with. In the previous iteration @enriquesanchez and I settled on making the labels for the font-size controls visible and instead hiding the "font size" fieldset legend:

<img width="561" alt="Screen Shot 2019-09-25 at 6 33 31 PM" src="https://user-images.githubusercontent.com/8096000/68270912-17671900-00b3-11ea-8c66-70acf6699b60.png">

tellthemachines reviewed (changes requested):
This is getting really close! 

I left some comments below, and besides that, there's a couple issues still needing attention. The dropdown is currently pushing the whole sidebar content up on desktop:
![font-picker-jump](https://user-images.githubusercontent.com/8096000/68271184-e0ddce00-00b3-11ea-9a35-6ab2fc3d7b64.gif)
This can be fixed by setting the overflow on `.components-panel` to `visible`, which doesn't seem to break anything else, but may need a little extra testing. It looks _really_ broken now though, so worth giving it a go 😅 

The other issue is with the `aria-labelledby` on the button. I'd mentioned previously that both VoiceOver and NVDA were reading out the button label repeatedly, and have now figured out that that is happening because `aria-labelledby` is pointing to the button's own id, and because the button text is overridden by its `aria-label` this means that essentially `aria-label` is read out twice.
<img width="585" alt="Screen Shot 2019-11-06 at 4 44 15 pm" src="https://user-images.githubusercontent.com/8096000/68271729-a412d680-00b5-11ea-9b3a-d598b5fbc003.png">
If we remove the reference to the button's id in that `aria-labelledby` the problem will be solved. This seems to be set in Downshift, is there any way to override it? Cc @silviuavram would be good to know why it was originally implemented in this way 🙂

silviuavram commented:
@tellthemachines yes, it comes from Downshift. I followed the pattern from [the ARIA website](https://www.w3.org/TR/wai-aria-practices/examples/listbox/listbox-collapsible.html).

It doesn't make sense to me to have a labelledby if the screen reader already knows how to read the text from it. It does not really help if the button has no text either.

I can add a PR in Downshift for this tomorrow, since I am at WebConference right now. Or you can do that and I will approve tomorrow. The labelledby should probably point only to the label. What do you think?

epiqueras committed:
Font Size Picker: A11y review.

epiqueras commented:
@tellthemachines Thanks for the thorough reviews and fixes :smile:! It should be good to go now.

>The labelledby should probably point only to the label. What do you think?

We don't need it in this case, because we explicitly set an `aria-label` which has better support from voice control software. It was double labelled, once by the `aria-label` and once by the `aria-labelledby`. I cleared the `aria-labelledby`.

silviuavram commented:
Oh ok, I understand. Interesting why a native button with text would ever need an aria-label. :(

epiqueras commented:
Voice control software doesn't follow the spec :cry:

enriquesanchez commented:
I did a quick test using Voice Control on the Mac:

![voiceControl 2019-11-15 17_53_55](https://user-images.githubusercontent.com/3276087/68983531-087a1800-07d1-11ea-8675-4aae97bfeaf7.gif)

[Here's the video recording](https://cldup.com/f0JbC7gh0m.m4v) 

Based on the little experience I've had with it, Voice Control is a bit different that Dragon. Manipulation is done via a grid and numbers, so labels have no effect here.

That being said, I was able to interact with the component and select different font sizes.

epiqueras commented:
Can we get another review please :smile:

cc @youknowriad @tellthemachines @gziolo

tellthemachines commented:
The labelling is behaving much better now! 
For clarity though, because "font size" is the fieldset legend that applies to all 3 fields inside it, each field should have its own label. That will help for Dragon users to be able to interact with them too. @enriquesanchez and I discussed this previously and came up with this design:

<img width="561" alt="Screen Shot 2019-09-25 at 6 33 31 PM" src="https://user-images.githubusercontent.com/8096000/69698382-48c68800-1139-11ea-9cea-ef97423390c1.png">

Where 
- the "font size" fieldset legend is visually hidden, because sighted users should be able to grasp what those fields are about from the surrounding context;
- "Size preset" (or "Preset size", which sounds better?) is the visible label for the picker, and
- "Custom" is the visible label for the range control.

There's also a little weirdness with the alignment of the options; the checked one is bumped slightly to the left due to the negative margin on the SVG. But that could be left for a follow-up PR 😅

epiqueras committed:
Font Size Picker: a11y iteration.

epiqueras commented:
@tellthemachines Thanks for taking another look! :smiley:

I think "Preset Size" sounds better. It makes more sense for "Size" to be the noun since its the more important word in this context.

I implemented that design and fixed the weirdness in the alignment with the check mark.

epiqueras committed:
Font Size Picker: Update tests.

epiqueras commented:
Just in time for the :turkey: :tada: !

youknowriad commented:
Nice work here :) A goodd follow-up would be to check where we're using `per_page: -1` in REST API calls and replace them with a paginated Select.

epiqueras commented:
>Nice work here :) A good follow-up would be to check where we're using per_page: -1 in REST API calls and replace them with a paginated Select.

It's on my list :smile:

youknowriad commented on packages/edit-post/src/components/sidebar/style.scss:
```diff
@@ -25,7 +26,7 @@
+			overflow: visible;
 			height: auto;
 			max-height: none;
 		}
```
Not sure I understand why the changes in this file are needed for this particular PR?

epiqueras replied:
https://github.com/WordPress/gutenberg/pull/17926#pullrequestreview-312188131

youknowriad replied:
Why it doesn't just use Popover, this issue is already solved there, and I think the changes here might cause more unexpected bugs (like the scrolling issue we noticed)

epiqueras replied:
Is the `Popover` fully accessible?

cc @tellthemachines @enriquesanchez

youknowriad replied:
If it's not it should be :) but we did a lot of a11y work on Popover back then.

enriquesanchez replied:
I'm not aware of any accessibility issues with `Popover`.

tellthemachines replied:
There are no issues that I know of with `Popover` itself, but we had some trouble with focus management when using it in our previous attempt to implement this component: https://github.com/WordPress/gutenberg/pull/17418#pullrequestreview-294158647

epiqueras replied:
Yes, that's what I was thinking of.

Part of the reason I didn't use `Popover` here from the start is that previous attempts had failed with it.

youknowriad commented on packages/components/src/index.js:
```diff
@@ -16,6 +16,7 @@ export { default as ClipboardButton } from './clipboard-button';
 export { default as ColorIndicator } from './color-indicator';
 export { default as ColorPalette } from './color-palette';
 export { default as ColorPicker } from './color-picker';
+export { default as CustomSelect } from './custom-select';
```
Why it's named CustomSelect and not just Select

epiqueras replied:
We already have `SelectControl` so that felt confusing, but I don't have a strong opinion on it.

youknowriad replied:
Good remark, I do think in that case that it should be SomethingControl. `CustomSelectControl` maybe. I don't like Custom that much though but I also don't have a lot of ideas.

epiqueras replied:
I prefer `Select` to be honest. The goal is for it to be THE React version of `select` so that makes sense.

`Control` makes it seem like it uses `BaseControl` which it doesn't.

tellthemachines replied:
I think it would be good to keep the "Custom", or some similar indication that this is not a standard HTML select (which we should also continue to provide, for cases where the options don't need to be visually distinguishable from one another)

youknowriad commented on storybook/test/__snapshots__/index.js.snap:
```diff
@@ -1539,61 +1539,78 @@ exports[`Storyshots Components|FontSizePicker Default 1`] = `
   className="components-font-size-picker"
 >
-  <legend>
+  <legend
```
useless snapshots that we change as the implementation is updated?

epiqueras replied:
Yes, from the "storyshots".

tellthemachines replied:
Why useless? I'd expect snapshots to change as the markup changes. I'd also expect to review those changes and make sure they are consistent with what we were trying to do, and nothing changed that shouldn't. That's the whole purpose of snapshots 🙂

youknowriad replied:
In my experience, each time you change the implementation, you have to change the snapshot and oftentimes you don't really check the details as it's a big blob of content.

Also, with snapshots it's not clear what are we testing exactly, we're not testing a specific behavior but just random content, it makes it hard to know what's a regression and what's not.

Another question: did you ever found aa regression with these snapshots?

epiqueras replied:
I think we use them for things like highlighting missing/removed class names that plugins rely on.

tellthemachines replied:
My take on it is snapshots are purely for testing markup: making sure no HTML has been accidentally stripped out or made unsemantic, classname or aria attribute changes, etc. It'll only be a big blob if we make a lot of changes! In this instance it's quite big because we changed the markup for the whole component. Still, I find markup changes much more readable in this format than in JSX.

I have found regressions before with snapshots; mostly of the kind mentioned above, but sometimes I've spotted issues with semantics on components being built from scratch too, because the snapshot view is pretty close to the output HTML.

Of course snapshots are only useful when we read them, but that is just a cultural issue that we can work on improving through educating on good testing and code review practice 😀

epiqueras replied:
Yeah, this is my take on them too.

youknowriad replied:
It's very hard to argue against tests :) and I had this exact same conversation over and over again with different people and everytime we land the snapshots and decide to rollback later :). Nothing personal here but I think this is just one of these occasions.

tellthemachines replied:
I'd love to know why you decided to roll back! In my previous project snapshots were pretty much universally agreed on. This PR might not be the best place for discussion 😄, but maybe we could bring it up at a Slack meeting at some point?

youknowriad replied:
What I said above. basically after some time using them, we noticed that they are not that useful.

youknowriad commented on packages/components/src/font-size-picker/index.js:
```diff
@@ -51,38 +57,40 @@ function FontSizePicker( {
+						label={ __( 'Preset Size' ) }
+						items={ items }
+						selectedItem={ items.find( ( item ) => item.key === currentSelectValue ) || items[ 0 ] }
+						onSelectedItemChange={ onSelectChangeValue }
```
Why not just onChange (I'm trying to see if we can mirror the SelectControl API props, so it's consistent and switching between the two becomes easy.

epiqueras replied:
I tried to mirror the `Downshift` props.

Their reasoning is probably that you also have `highlightedItem`, `onHighlightedItemChange`, etc. They prefer to be explicit about the two instead of having `selectedItem == value`.

youknowriad replied:
I have a preference for staying consistent across our components, Downshift is just an implementation detail that can be changed at any time.

epiqueras replied:
I think that having `onChange` and `onHighlightedValueChange` is a bit confusing. It's not totally clear that `onChange` does not include highlighted value changes.

Also, `selectedItem` takes the whole object, not just the key.

Similarly, `onSelectedItemChange` gives you the whole object, not just the key.

This already makes them fundamentally different to `SelectControl`'s `value` and `onChange`, but they are a much simpler to use API as you don't have to handle all of the key-to-object normalization/searching on the consumer side.

youknowriad replied:
I feel like highlighting is not something you'd need to care about often as a consumer, I'm not sure we even need it as an API in our component to be honest.

Also, I just noticed that the component is lacking a README.

epiqueras replied:
We already have the functionality for free and this is a reusable component meant to be used outside of Gutenberg as well so I don't see how we gain from not exposing it.

Also, `selectedItem` and `items` very clearly convey the contract that the `selectedItem` should be an object from the `items` list.

>Also, I just noticed that the component is lacking a README.

I will work on one, but are you still unhappy with the prop names? What props would you expose?

youknowriad replied:
> Also, selectedItem and items very clearly convey the contract that the selectedItem should be an object from the items list.

I get that but the boat has sailed for a lot of components now, and I'd rather have users learn one way to use all of our controls rather than having each component with its own specifics.

> I will work on one, but are you still unhappy with the prop names? What props would you expose?

value, onChange and options being the important ones. 

>  I don't see how we gain from not exposing it.

we gain that if we ever change the implementation or anything, we're less likely to break backward compatibility. Adding new APIs anywhere in our codebase shouldn't be considered a "good thing", it's something to do with a lot of caution.

epiqueras replied:
>I get that but the boat has sailed for a lot of components now, and I'd rather have users learn one way to use all of our controls rather than having each component with its own specifics.

That's fair and I can definitely see value in that if we stick with it across all inputs. I updated the prop names in this PR: #18827.

>we gain that if we ever change the implementation or anything, we're less likely to break backward compatibility. Adding new APIs anywhere in our codebase shouldn't be considered a "good thing", it's something to do with a lot of caution.

Agreed, but in this case, it's something all select implementations could easily expose, so I wouldn't worry about backwards compatibility with it. I still haven't exposed it for now to be safe, we can do it when the need for it arises.

afercia commented:
Looks like there are a few missing things in this component, will comment on the issue.

aduth commented:
Noting that this Downshift dependency is fairly large at 32.4kb, contributing ~200ms additional load time over 3G.

https://bundlephobia.com/result?p=downshift@3.4.5

silviuavram commented:
It is true, the bundle size is getting bigger, and I am close to release `useCombobox`. I am considering having useSelect and other hooks as part of a a11y monorepo, as this size issue popped up to us recently as well. If you'd like to help out in converging to a solution to the bundle size that would be fantastic.

It should not break the current behavior (`useSelect` being as part of `Downshift`) and also it should be part of a monorepo with a11y hooks. It would be interesting if it can be imported in Downshift from that monorepo and tree-shaken if not used. I am pretty new to these concepts so I would love some opinions. Maybe @kentcdodds can help out with an idea.

Thoughts?

epiqueras commented:
>It should not break the current behavior (useSelect being as part of Downshift) and also it should be part of a monorepo with a11y hooks. It would be interesting if it can be imported in Downshift from that monorepo and tree-shaken if not used.

Sounds like a reasonable plan.

silviuavram commented:
Although I am thinking ... if you are using `import` to get only `useSelect`, and I am using `rollup` to bundle, shouldn't you get only that hook, and not the whole library?

epiqueras commented:
Yeah `Downshift` has an `esm` distribution. I think our tooling doesn't support it, but we can work on that right @aduth?

aduth commented:
If it can be tree-shaken, our build should be able to drop the unused code. I was judging based on the [BundlePhobia page](https://bundlephobia.com/result?p=downshift@3.4.5), specifically this text under "Exports Analysis":

>Exports analysis is available only for packages that export ES Modules and are side-effect free. This package exports ES6 modules, but isn't marked side-effect free.

epiqueras commented:
I guess it's not explicitly marked side-effect free, but I don't think it has any side-effects does it?

aduth commented:
It still needs to be included in the published `package.json` for Webpack to consider it for tree shaking:

>The new webpack 4 release expands on this capability with a way to provide hints to the compiler via the "`sideEffects`" `package.json` property to denote which files in your project are "pure" and therefore safe to prune if unused.

https://webpack.js.org/guides/tree-shaking/#mark-the-file-as-side-effect-free

epiqueras commented:
It looks like it will still try to tree-shake most things, but will trip up with things like HOCs without directives:

https://webpack.js.org/guides/tree-shaking/#clarifying-tree-shaking-and-sideeffects

@silviuavram can we add the `sideEffects` hint to make this more efficient?

silviuavram commented:
I am really interested to try this out, I will make changes on `Downshift` side if needed. Let me know on the progress please!

epiqueras commented:
@silviuavram it's `Downshift` that is missing the `sideEffects` property, we need to add it there.

silviuavram commented:
Can you create a PR and I will look over it tomorrow? Thank you! Do we need
to publish an alpha version to test if it works?

On Wed, 4 Dec 2019 at 18:19, Enrique Piqueras <notifications@github.com>
wrote:

> @silviuavram <https://github.com/silviuavram> it's Downshift that is
> missing the sideEffects property, we need to add it there.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=ACWAZAEBES6R3WBPJB3BDA3QW7RCXA5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF5YV4I#issuecomment-561744625>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWAZAEFWUKASHJVNY2NJW3QW7RCXANCNFSM4JAIMOLQ>
> .
>

epiqueras commented:
@silviuavram https://github.com/downshift-js/downshift/pull/843

>Do we need
to publish an alpha version to test if it works?

Yeah, that would be helpful. It should work, unless `Downshift` has side effects, but better to be safe.

silviuavram commented:
Thanks very much! I will publish an alpha tomorrow from that branch and
test it out.

On Wed, 4 Dec 2019 at 18:35, Enrique Piqueras <notifications@github.com>
wrote:

> @silviuavram <https://github.com/silviuavram> downshift-js/downshift#843
> <https://github.com/downshift-js/downshift/pull/843>
>
> Do we need
> to publish an alpha version to test if it works?
>
> Yeah, that would be helpful. It should work, unless Downshift has side
> effects, but better to be safe.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=ACWAZACDGJCFX4WO4ZEKWXDQW7S6XA5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEF533UQ#issuecomment-561757650>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWAZADXQ3NX46KJ3SOKF23QW7S6XANCNFSM4JAIMOLQ>
> .
>

silviuavram commented:
I published downshift@3.5.0-alpha.0 and could not manage to properly test it. Maybe my testing project is not configured as it should. If you have any tips that will help me looking into it tomorrow that will be super helpful.

epiqueras commented:
What was the issue?

On Thu, Dec 5, 2019 at 12:15 PM Silviu Alexandru Avram <
notifications@github.com> wrote:

> I published downshift@3.5.0-alpha.0 and could not manage to properly test
> it. Maybe my testing project is not configured as it should. If you have
> any tips that will help me looking into it tomorrow that will be super
> helpful.
>
> —
> You are receiving this because you modified the open/close state.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=AESFA2GOTNEADODBACKASW3QXFONDA5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGB76SQ#issuecomment-562298698>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/AESFA2EGR6DXCORG43C43ETQXFONDANCNFSM4JAIMOLQ>
> .
>

silviuavram commented:
I probably did not know what to look for in the minified file. In the
main.map.js I could still find all the exports (useSelect, Downshift,
useCombobox) even though I only imported: import {useCombobox} from
‘downshift’. I probably did not configure the webpack in my test project
correctly. Or it’s something wrong in Downshift’s rollup config or
index.js. Or maybe it works and I could not figure tjat out from the
minified file. I don’t know yet.

On Thu, 5 Dec 2019 at 21:55, Enrique Piqueras <notifications@github.com>
wrote:

> What was the issue?
>
> On Thu, Dec 5, 2019 at 12:15 PM Silviu Alexandru Avram <
> notifications@github.com> wrote:
>
> > I published downshift@3.5.0-alpha.0 and could not manage to properly
> test
> > it. Maybe my testing project is not configured as it should. If you have
> > any tips that will help me looking into it tomorrow that will be super
> > helpful.
> >
> > —
> > You are receiving this because you modified the open/close state.
> > Reply to this email directly, view it on GitHub
> > <
> https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=AESFA2GOTNEADODBACKASW3QXFONDA5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGB76SQ#issuecomment-562298698
> >,
> > or unsubscribe
> > <
> https://github.com/notifications/unsubscribe-auth/AESFA2EGR6DXCORG43C43ETQXFONDANCNFSM4JAIMOLQ
> >
> > .
> >
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=ACWAZABWTCLZN6KD6HMHY63QXFTF5A5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGCDWII#issuecomment-562314017>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWAZAGQVQCTX7ZLIGTEZ5TQXFTF5ANCNFSM4JAIMOLQ>
> .
>

epiqueras commented:
Can you push your test somewhere?

silviuavram commented:
https://github.com/silviuavram/check-downshift-shake tried it here.

silviuavram commented:
I updated the repo with some things I tried on babel. Still it imports the whole things in the ES bundle.

epiqueras commented:
Yeah, you're building in development mode:

https://github.com/silviuavram/check-downshift-shake/blob/27a5fcb763ce2b837017a948a6c859ccfa0a0148/package.json#L11-L12

See, https://webpack.js.org/guides/tree-shaking/.

silviuavram commented:
I tried to build with —production as well. The es has the same size as the
cjs.

On Fri, 6 Dec 2019 at 18:40, Enrique Piqueras <notifications@github.com>
wrote:

> Yeah, you're building in development mode:
>
>
> https://github.com/silviuavram/check-downshift-shake/blob/27a5fcb763ce2b837017a948a6c859ccfa0a0148/package.json#L11-L12
>
> See, https://webpack.js.org/guides/tree-shaking/.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/WordPress/gutenberg/pull/17926?email_source=notifications&email_token=ACWAZAGDSWVR44R3RIMB25TQXKE7ZA5CNFSM4JAIMOL2YY3PNVWWK3TUL52HS4DFVREXG43VMVBW63LNMVXHJKTDN5WW2ZLOORPWSZGOEGE2LHQ#issuecomment-562668958>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ACWAZAEL6VZESKP6KJFGAKTQXKE7ZANCNFSM4JAIMOLQ>
> .
>

epiqueras commented:
I just debugged this and it's probably because of how `Downshift` bundles its `esm` module.

There are a lot of static property assignments in there like `Downshift.stateChangeTypes = stateChangeTypes;` and `useCombobox.stateChangeTypes = stateChangeTypes$2;`.

Some prior discussion:

https://github.com/downshift-js/downshift/issues/380

silviuavram commented:
Looks good, I was also suspicious about those state change things while going through the bundled code. Thanks so much for looking into this!

The bad part is the fact that we need to deliver a breaking change to make the bundle tree-shakeable again: exporting those `stateChangeTypes` separately and removing the static allocation. Will check if indeed, with these changes, the problem will be fixed.

epiqueras commented:
There are other workarounds:

- You could break down the output into multiple files with their ESM imports intact and that would allow the `sideEffects` flag to work. Currently, Webpack evaluates the single output file, because there are used exports there, and it can't prune the unused ones because they are mutated. If they were imported and mutated from a different file, Webpack could use the `sideEffects` hint and avoid evaluating those files.

- Another, perhaps easier change, would be to use the `/*#__PURE__*/` annotations. They're like the `sideEffects` hint, but for statements instead of modules.

Related discussion: https://github.com/styled-components/babel-plugin-styled-components/issues/245.

silviuavram commented:
Found a few problems with today's investigation:

The hooks were not tree-shaken because of
```js
const validatePropTypes = getPropTypesValidator(useSelect, propTypes)
```

`Downshift` component not tree-shaken because of the `static` props `stateChangeTypes` and `defaultProps!!!`. The `propTypes` static one was correctly removed by a babel transpiler (I think) at bundle time, like:
```process.env.NODE_ENV !== "production" ? Downshift.propTypes = {```

Probably I could add this same check manually to the hooks, because I don't need that prop types check function to run in production. On the `Downshift` static fields `stateChangeTypes` and `defaultProps` maybe there is a solution with one of your suggestions?

epiqueras commented:
Yeah, both should work, albeit the first one is a better long-term solution that will work for future code as well.

youknowriad commented:
Any plans to add a README for this component.

epiqueras commented:
#19026

adamsilverstein commented:
With Downshift merged, we should reconsider using it elsewhere (eg for Author selection, page parent selection, etc... eg https://github.com/WordPress/gutenberg/pull/7478). 

@afercia I missed most of the work/conversation here - does the progress on Downshift address the feedback from the original ticket? shall I work on a new PR for you to review?

epiqueras commented:
Yes, that is a good next step with it.
